lib/puppet/provider/check_http/check_http.rb

```diff
@@ -1,21 +1,60 @@
 # frozen_string_literal: true
 
+require 'puppet/resource_api'
+require 'puppet/resource_api/simple_provider'
 require 'net/http'
+require 'retriable'
 
 # Implementation for the check_http type using the Resource API.
 class Puppet::Provider::CheckHttp::CheckHttp
   def get(_context)
     []
   end
 
-  def set(context, changes)
-    changes.each do |name, _change|
-      uri = URI(name)
-      context.processing(uri.to_s, {}, {}, message: 'checking http') do
-        if Net::HTTP.get(uri)
-          context.info("successfully connected to #{name}")
+  def set(context, changes); end
+
+  # Update the check_http provider to use the above attributes to execute up to retries number of times
+  # with success being defined as having one of the expected_statuses
+  # and the body of the response matches body_matcher while taking into account request_timeout.
+
+  def insync?(context, name, attribute_name, _is_hash, should_hash)
+    context.debug("Checking whether #{attribute_name} is up-to-date")
+    uri = URI.parse(should_hash[:url])
+
+    # Update the check_http provider to wait for backoff ** (exponential_backoff_base * (retries - 1) seconds between attempts (up to max_backoff)
+    base_interval = should_hash[:backoff]**(should_hash[:exponential_backoff_base] * should_hash[:retries] - 1)
+
+    # This callback provides the exception that was raised in the current try, the try_number, the elapsed_time for all tries so far, and the time in seconds of the next_interval.
+    do_this_on_each_retry = proc do |exception, try, elapsed_time, next_interval|
+      context.info("#{exception.class}: '#{exception.message}' - #{try} tries in #{elapsed_time} seconds and #{next_interval} seconds until the next try.") unless exception.nil?
+    end
+
+    base_interval = 0
+    retried = 0
+    Retriable.retriable(tries: should_hash[:retries], max_elapsed_time: should_hash[:request_timeout], base_interval: base_interval, max_interval: should_hash[:max_backoff],
+multiplier: should_hash[:exponential_backoff_base], on_retry: do_this_on_each_retry) do
+      retried += 1
+      # Update the check_http provider to wait for backoff ** (exponential_backoff_base * (retries - 1) seconds between attempts (up to max_backoff)
+      base_interval = should_hash[:backoff]**should_hash[:exponential_backoff_base] * (retried - 1)
+      context.info("The base_interval is '#{base_interval}' and retrying for '#{retried}' time")
+      response = Net::HTTP.get_response(uri)
+
+      # Success being defined as having one of the expected_statuses and the body of the response matches body_matcher
+      unless response.nil?
```

Comment:
can `response` ever be `nil`? and if so, shouldn't this also raise an error then?

Reply:
Removed response.nil condition since the value will be nil only when the error is raised and its handled already  here in Retriable
https://github.com/puppetlabs/opv/pull/13/files#diff-37480ed32724bc6fb660379989da375be68ebd34ba5083aeadd4b4d5c21b9b79R25